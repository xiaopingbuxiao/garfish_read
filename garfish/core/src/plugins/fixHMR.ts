import { interfaces } from '../interface';

// When the main application is updated, the currently active child applications need to rerender.
export function GarfishHMRPlugin() {
  let hasInit = false;
  let isHotUpdate = false;
  return function (Garfish: interfaces.Garfish): interfaces.Plugin {
    return {
      name: 'fix-hmr',
      version: __VERSION__,

      bootstrap() {
        if (hasInit) return;
        hasInit = true;

        let webpackHotUpdateName = 'webpackHotUpdate';
        let webpackHotUpdate = (window as any)[webpackHotUpdateName];

        // 查找 webpackHotUpdate 函数
        for (const i in window) {
          if (i.includes('webpackHotUpdate')) {
            webpackHotUpdateName = i;
            webpackHotUpdate = window[i];
          }
        }

        if (typeof webpackHotUpdate === 'function') {
          (window as any)[webpackHotUpdateName] = function () {
            isHotUpdate = true;
            return webpackHotUpdate.apply(this, arguments);
          };

          const observer = new MutationObserver(() => {
            if (!isHotUpdate) return;
            isHotUpdate = false;

            Garfish.activeApps.forEach((app) => {
              if (app.mounted) {
                setTimeout(() => {
                  app.display && app.hide();
                  app.show();
                });
              }
            });
          });

          observer.observe(document.documentElement, {
            subtree: true,
            childList: true,
            attributes: true,
          });
        }
      },
    };
  };
}
