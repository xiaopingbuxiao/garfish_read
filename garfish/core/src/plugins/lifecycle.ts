import { interfaces } from '../interface';

export function GarfishOptionsLife(options, name: string) {
  return function (): interfaces.Plugin {
    return {
      name,
      version: __VERSION__,
      ...options,
    };
  };
}
