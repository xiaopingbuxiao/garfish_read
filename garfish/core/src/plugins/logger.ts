import { coreLog } from '@garfish/utils';
import { interfaces } from '../interface';

export function GarfishLogger() {
  return function (): interfaces.Plugin {
    return {
      name: 'garfish-logger',
      version: __VERSION__,
      beforeLoad(appInfo, ...args) {
        coreLog(`${appInfo.name} beforeLoad`, [appInfo, ...args]);
      },
      afterLoad(appInfo, appInstance, ...args) {
        if (appInstance) {
          coreLog(`${appInfo.name} id: ${appInstance.appId} afterLoad`, [
            appInfo,
            ...args,
          ]);
        }
      },
      beforeMount(appInfo, appInstance, ...args) {
        coreLog(`${appInfo.name} id: ${appInstance.appId} beforeMount`, [
          appInfo,
          ...args,
        ]);
      },
      afterMount(appInfo, appInstance, ...args) {
        coreLog(`${appInfo.name} id: ${appInstance.appId} afterMount`, [
          appInfo,
          ...args,
        ]);
      },
      beforeUnmount(appInfo, appInstance, ...args) {
        coreLog(`${appInfo.name} id: ${appInstance.appId} beforeUnmount`, [
          appInfo,
          ...args,
        ]);
      },
      afterUnmount(appInfo, appInstance, ...args) {
        coreLog(`${appInfo.name} id: ${appInstance.appId} afterUnmount`, [
          appInfo,
          ...args,
        ]);
      },
    };
  };
}
