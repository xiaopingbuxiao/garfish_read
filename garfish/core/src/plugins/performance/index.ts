import { getRenderNode } from '@garfish/utils';
import { interfaces } from '../../index';
import { SubAppObserver } from './subAppObserver';

// Key nodes in Garfish corresponding to the life cycle of registration
export function GarfishPerformance() {
  return function (): interfaces.Plugin {
    const subAppMap = {};
    return {
      name: 'performance',

      beforeLoad(appInfo) {
        if (!subAppMap[appInfo.name] && appInfo.domGetter) {
          subAppMap[appInfo.name] = new SubAppObserver({
            subAppRootSelector: appInfo.domGetter,
          });
        }
        subAppMap[appInfo.name].subAppBeforeLoad(appInfo.entry);
      },

      afterLoad(appInfo, appInstance: interfaces.App) {
        if (appInstance) {
          appInstance.appPerformance = subAppMap[appInfo.name] as any;
        }
      },

      beforeMount(appInfo) {
        subAppMap[appInfo.name].subAppBeforeMount(appInfo.entry);
      },

      beforeUnmount(appInfo) {
        subAppMap[appInfo.name].subAppUnmount(appInfo.entry);
      },
    };
  };
}
