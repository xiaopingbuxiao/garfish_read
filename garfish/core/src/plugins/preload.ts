import {
  warn,
  Queue,
  isAbsolute,
  transformUrl,
  idleCallback,
  callTestCallback,
  safeWrapper,
} from '@garfish/utils';
import { Loader, Manager, TemplateManager } from '@garfish/loader';
import { interfaces } from '../interface';

export const storageKey = '__garfishPreloadApp__';

const isMobile =
  /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
    navigator.userAgent,
  );

// Using queues, to avoid interference with the normal request
const requestQueue = new Queue();

const isSlowNetwork = () =>
  (navigator as any).connection
    ? (navigator as any).connection.saveData ||
      /(2|3)g/.test((navigator as any).connection.effectiveType)
    : false;

export const requestIdleCallback =
  __TEST__ || typeof idleCallback !== 'function'
    ? window.setTimeout
    : idleCallback;

// Test size, catch mistakes, avoid preload first screen white during parsing error
function safeLoad({
  loader,
  appName,
  url,
  isModule,
  immediately,
  callback,
}: {
  loader: Loader;
  appName: string;
  url: string;
  isModule: boolean;
  immediately: boolean;
  callback?: (m: Manager) => any;
}) {
  const generateSuccess =
    (next: () => void = () => {}) =>
    ({ resourceManager }) => {
      callback && callback(resourceManager);
      setTimeout(next, 500);
    };

  const generateThrowWarn =
    (next: () => void = () => {}) =>
    (e) => {
      if (__DEV__) {
        warn(e);
        warn(`Preload failed. "${url}"`);
      }
      next();
    };

  const loadResource = (next: () => void = () => {}) => {
    const throwWarn = generateThrowWarn(next);
    const success = generateSuccess(next);
    try {
      if (isModule) {
        loader.loadModule(url).then(success, throwWarn);
      } else {
        loader.load({ scope: appName, url }).then(success, throwWarn);
      }
    } catch (e) {
      throwWarn(e);
      next();
    }
  };

  if (immediately) {
    loadResource();
  } else {
    requestQueue.add((next) => {
      requestIdleCallback(() => loadResource(next));
    });
  }
}

export function loadAppResource(
  loader: Loader,
  info: interfaces.AppInfo,
  immediately = false,
) {
  __TEST__ && callTestCallback(loadAppResource, info);
  const fetchUrl = transformUrl(location.href, info.entry);

  safeLoad({
    loader,
    appName: info.name,
    url: fetchUrl,
    isModule: false,
    immediately,
    callback: (manager) => {
      const loadStaticResource = () => {
        if (manager instanceof TemplateManager) {
          const baseUrl = manager.url;
          const jsNodes = manager.findAllJsNodes();
          const linkNodes = manager.findAllLinkNodes();
          const metaNodes = manager.findAllMetaNodes();

          if (jsNodes) {
            jsNodes.forEach((node) => {
              const src = manager.findAttributeValue(node, 'src');
              src &&
                safeLoad({
                  loader,
                  appName: info.name,
                  url: baseUrl ? transformUrl(baseUrl, src) : src,
                  isModule: false,
                  immediately,
                });
            });
          }
          if (linkNodes) {
            linkNodes.forEach((node) => {
              if (manager.DOMApis.isCssLinkNode(node)) {
                const href = manager.findAttributeValue(node, 'href');
                href &&
                  safeLoad({
                    loader,
                    appName: info.name,
                    url: baseUrl ? transformUrl(baseUrl, href) : href,
                    isModule: false,
                    immediately,
                  });
              }
            });
          }
          if (metaNodes) {
            metaNodes.forEach((node) => {
              if (manager.DOMApis.isRemoteModule(node)) {
                const src = manager.findAttributeValue(node, 'src');
                if (src && isAbsolute(src)) {
                  safeLoad({
                    loader,
                    appName: info.name,
                    url: src,
                    isModule: true,
                    immediately,
                  });
                } else {
                  warn(
                    `The loading of the remote module must be an absolute path. "${src}"`,
                  );
                }
              }
            });
          }
        }
      };
      if (immediately) {
        loadStaticResource();
      } else {
        requestIdleCallback(loadStaticResource);
      }
    },
  });
}

export function getRanking() {
  const str = localStorage.getItem(storageKey);
  if (str) {
    const data = JSON.parse(str);
    return data.sort((a, b) => b.count - a.count);
  }
  return [];
}

export function setRanking(appName: string) {
  const str = localStorage.getItem(storageKey);
  const newCurrent = { appName, count: 1 };

  if (!str) {
    safeWrapper(() =>
      localStorage.setItem(storageKey, JSON.stringify([newCurrent])),
    );
  } else {
    const data = JSON.parse(str);
    const current = data.find((app) => app.appName === appName);
    current ? current.count++ : data.push(newCurrent);
    safeWrapper(() => localStorage.setItem(storageKey, JSON.stringify(data)));
  }
}

const loadedMap = Object.create(null); // Global cache, only load again is enough

declare module '@garfish/core' {
  export default interface Garfish {
    preloadApp: (appName: string) => void;
  }
}

export function GarfishPreloadPlugin() {
  return function (Garfish: interfaces.Garfish): interfaces.Plugin {
    Garfish.preloadApp = (appName) => {
      loadAppResource(Garfish.loader, Garfish.appInfos[appName], true);
    };

    return {
      name: 'preload',
      version: __VERSION__,

      beforeLoad(appInfo) {
        if (Garfish.options.disablePreloadApp) {
          return;
        }
        setRanking(appInfo.name);
      },

      registerApp(appInfos) {
        // Through disablePreloadApp preload is prohibited
        if (Garfish.options.disablePreloadApp) {
          return;
        }
        setTimeout(
          () => {
            if (isMobile || isSlowNetwork()) return;
            const ranking = getRanking();

            for (const { appName } of ranking) {
              if (appInfos[appName] && !loadedMap[appName]) {
                loadedMap[appName] = true;
                loadAppResource(Garfish.loader, appInfos[appName]);
              }
            }

            for (const key in appInfos) {
              if (!loadedMap[key]) {
                loadAppResource(Garfish.loader, appInfos[key]);
              }
            }
          },
          __TEST__ ? 0 : 5000,
        );
      },
    };
  };
}
