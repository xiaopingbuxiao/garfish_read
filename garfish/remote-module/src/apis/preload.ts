import { assert, isAbsolute } from '@garfish/utils';
import { hooks } from '../hooks';
import { processAlias } from './setModuleConfig';
import { loader, resourcesStore } from '../common';

// Preload the static resources of the module, so that the module can be loaded synchronously
export function preload(urls: string | Array<string>) {
  if (!Array.isArray(urls)) urls = [urls];

  return Promise.all(
    urls.map((url) => {
      url = processAlias(url)[0];
      assert(
        isAbsolute(url),
        `The loading of the remote module must be an absolute path. "${url}"`,
      );
      return loader.loadModule(url).then((data) => {
        if (data.resourceManager) {
          data.resourceManager.originUrl = url;
          resourcesStore.push(data.resourceManager);
          hooks.lifecycle.preloaded.emit(data.resourceManager);
        }
        return data;
      });
    }),
  );
}
