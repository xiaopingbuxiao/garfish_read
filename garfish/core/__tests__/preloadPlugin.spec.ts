import { mockStaticServer } from '@garfish/test-suite';
import Garfish from '../src/index';
import { GarfishPreloadPlugin, storageKey } from '../src/plugins/preload';

describe('Core: preload plugin', () => {
  const vueSubAppEntry = './resources/vueApp.html';
  const reactSubAppEntry = './resources/reactApp.html';

  mockStaticServer({
    baseDir: __dirname,
  });

  it('disablePreloadApp is true close setRanking', async () => {
    const GarfishInstance = new Garfish({});
    GarfishInstance.run({
      disablePreloadApp: true,
      apps: [
        {
          name: 'vue-app',
          entry: vueSubAppEntry,
        },
      ],
    });
    await GarfishInstance.loadApp('vue-app');
    expect(localStorage.getItem(storageKey)).toBeNull();
  });

  it('disablePreloadApp is false use setRanking', async () => {
    const GarfishInstance = new Garfish({});
    GarfishInstance.run({
      disablePreloadApp: false,
      apps: [
        {
          name: 'vue-app',
          entry: vueSubAppEntry,
        },
        {
          name: 'react-app',
          entry: reactSubAppEntry,
        },
      ],
    });
    await GarfishInstance.loadApp('vue-app');
    await GarfishInstance.loadApp('vue-app');
    await GarfishInstance.loadApp('react-app');
    expect(localStorage.getItem(storageKey)).toMatchSnapshot();
  });
});
