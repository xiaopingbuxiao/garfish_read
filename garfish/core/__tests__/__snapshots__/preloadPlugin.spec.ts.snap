// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core: preload plugin disablePreloadApp is false use setRanking 1`] = `"[{\\"appName\\":\\"vue-app\\",\\"count\\":2},{\\"appName\\":\\"react-app\\",\\"count\\":1}]"`;
